### ISSUE TITLE
Make mutate function return a Future

### ISSUE THREAD

**Issue by smkhalsa** [zino-app/graphql-flutter#273]
the mutate function currently returns `void`. I suggest we change it to return a Future with the mutation result.

**Comment by smkhalsa** [comment]
In case anyone else lands here, an alternative is to call the mutate function on the `GraphqlClient`'s `QueryManager` directly. I am now using `package:provider` to pass my `GraphqlClient` down my widget tree. This approach is very similar to React's hooks, and I actually prefer it to the Mutation widget.

**Comment by mainawycliffe** [comment]
Yea, i agree with you in regards to mutations, should definitely return a future with the results from the server. If you prefer not to use the widgets, there is a separate core package, which you can use with any state management solution of your choice. Though its still in beta. You can learn more [here](https://github.com/zino-app/graphql-flutter/blob/beta/packages/graphql)

**Comment by smkhalsa** [comment]
@mainawycliffe Looking through the source, it looks like the mutate function in the graphql package doesn't accept an `update` function to update the cache. I can open a separate issue for that. That would be important to have for my use case.

**Comment by mainawycliffe** [comment]
Yes, open a new issue, i will pick it once am done with the current feature am working on or someone else will.

**Comment by micimize** [comment]
This would indeed be [more in line with Apollo](https://www.apollographql.com/docs/react/essentials/mutations#render-prop):
>  The function returns a promise that fulfills with your mutation result.

But there is a question as to whether it should fulfill optimistically

**Comment by smkhalsa** [comment]
@micimize good question.

In the case of Apollo, the render prop function provides a `mutate` function and a `MutationResponse` object. I believe the `mutate` function returns a Promise once the client receives a server response, and the `MutationResponse` updates immediately with the optimistic response. However, I haven't dug into the source to validate.

The above approach makes sense to me. Since optimistic responses resolve immediately and you already know what the result is, it doesn't seem very useful to have the `mutate` Future resolve with an optimistic result.

**Comment by mainawycliffe** [comment]
I am looking at the code and mutate functions already return the future. It behaves and just like the Query method behaves for the most part.

**Comment by smkhalsa** [comment]
@mainawycliffe it looks like the mutation on QueryManager does return a future, but the [runMutation](https://github.com/zino-app/graphql-flutter/blob/master/packages/graphql_flutter/lib/src/widgets/mutation.dart#L8) function in the Mutation Widget returns void.

Seems like this should be an easy fix.

**Comment by mainawycliffe** [comment]
@smkhalsa actually i don't think it needs a fix in this case. Since this the widget the results of a mutation are received by the parent widget of the mutation widget via `StreamBuilder` in the same way the Query widget does.

**Comment by smkhalsa** [comment]
So here's a use case: I have a list of posts and an overlay textinput widget to add a new post. When I submit a new post, I close the overlay and using the update function add the new post to the list. Once the mutation Future returns successfully, I want to show a snackbar at the bottom of the screen confirming that my post was successfully added.

**Comment by mainawycliffe** [comment]
I am spit-balling here but i would look at the `onCompleted` and `update` method exposed by the widget. You can pass a callback to the `onCompleted` method and it will be called once the mutation is complete or dispatch a state change (this would work with any state management solution - bloc, redux) indicating mutation is complete.

**Comment by smkhalsa** [comment]
But why wouldn't we want to Mutate function to return a future? Just in terms of organization, dealing with the side effects based on the response of a mutation in the same place you're calling the function just makes more sense.

**Comment by mainawycliffe** [comment]
Am not opposed to that. The concern i have is if we convert `runMutation` into a `Future<QueryResults>` defeats the purpose of using a widget. You would use the Mutation Widget, then you would have to manually subscribe to the returned future, this is kind of repetitive. If this is the use case, you can call the mutation method directly, without the [mutation method](https://github.com/zino-app/graphql-flutter/blob/0c176adfb8f8eb4b302db0fb591aac23511c496f/packages/graphql/example/bin/main.dart#L79), and you will get a future.

**Comment by micimize** [comment]
This is tough - it does kind of feel like apollo is making a mistake here
<details>
<summary>I'm pretty sure their Promise is optimistic and binding logic to this api surface area could result in unexpected results
</summary>
<br/>

Apollo usually doesn't externalize the optimistic state, and just pretends `optimisticResults` are real results. `runMutation` pretty transparently passes down to [`QueryManager.mutate`](https://github.com/apollographql/apollo-client/blob/master/packages/apollo-client/src/core/QueryManager.ts#L132), which will then `resolve(storeResult!);`, so that future is optimistic when optimism is to be had
</details>

Using `onCompleted` is definitely more idiomatic. We could also just return a `Stream` wrapper with `.eager`,  and `.fromNetwork` or something, which I think if we implement this would be the way to go.

**Comment by joprice** [comment]
On the same vein @smkhalsa's comments, some widgets use `Future`s for controlling their state, like `Dismissible`, which has `confirmDismiss`. I tried using the client directly to work around this, and then after digging into the code, realized that this doesn't run the same cache updating and rebroadcasting logic as the `Mutation` widget, so I tried creating a `Completer` instead, which I fulfill in the `update` callback. This seems to work so far, but it seems like having a `Future` response from `RunMutation` would make it simpler.
